cmd/botkube/main.go

```diff
@@ -56,97 +57,156 @@ const (
+		reportFatalErr(logger, reporter, wrappedErr)
+		return wrappedErr
+	}
+
+	// Register current anonymous identity
+	k8sCli, err := kubernetes.NewForConfig(kubeConfig)
+	if err != nil {
+		wrappedErr := fmt.Errorf("while creating K8s clientset: %w", err)
+		reportFatalErr(logger, reporter, wrappedErr)
+		return wrappedErr
+	}
+	err = reporter.RegisterCurrentIdentity(ctx, k8sCli, appCfg.ConfigPath)
+	if err != nil {
+		wrappedErr := fmt.Errorf("while registering current identity: %w", err)
+		reportFatalErr(logger, reporter, wrappedErr)
+		return wrappedErr
+	}
+
 	// Prometheus metrics
 	metricsSrv := newMetricsServer(logger.WithField(componentLogFieldKey, "Metrics server"), appCfg.MetricsPort)
 	errGroup.Go(func() error {
+		defer analytics.ReportPanicIfOccurs(logger, reporter)
 		return metricsSrv.Serve(ctx)
 	})
 
+	// Load config
 	conf, err := config.Load(appCfg.ConfigPath)
-	exitOnError(err, "while loading configuration")
+	if err != nil {
+		wrappedErr := fmt.Errorf("while loading configuration: %w", err)
+		reportFatalErr(logger, reporter, wrappedErr)
+		return wrappedErr
+	}
 	if conf == nil {
-		log.Fatal("while loading configuration: config cannot be nil")
+		wrappedErr := fmt.Errorf("while loading configuration: %w", errors.New("config cannot be nil"))
+		reportFatalErr(logger, reporter, wrappedErr)
+		return wrappedErr
 	}
 
-	// Prepare K8s clients and mapper
-	kubeConfig, err := clientcmd.BuildConfigFromFlags("", appCfg.KubeconfigPath)
-	exitOnError(err, "while loading K8s config")
-
-	dynamicCli, discoveryCli, mapper, err := getK8sClients(kubeConfig)
-	exitOnError(err, "while getting K8s clients")
-
-	// Set up analytics reporter
-	analyticsReporter, cleanupFn, err := newAnalyticsReporter(appCfg.Analytics.Disable, logger)
-	exitOnError(err, "while creating analytics reporter")
-	defer cleanupFn()
-
-	k8sCli, err := kubernetes.NewForConfig(kubeConfig)
-	exitOnError(err, "while creating K8s clientset")
-	err = analyticsReporter.RegisterCurrentIdentity(ctx, k8sCli, appCfg.ConfigPath)
-	exitOnError(err, "while registering current identity")
-
 	// Set up the filter engine
 	filterEngine := filterengine.WithAllFilters(logger, dynamicCli, mapper, conf)
 
 	// List notifiers
-	notifiers, err := notify.LoadNotifiers(logger, conf.Communications)
-	exitOnError(err, "while loading notifiers")
+	notifiers, err := notify.LoadNotifiers(logger, conf.Communications, reporter)
+	if err != nil {
+		wrappedErr := fmt.Errorf("while loading notifiers: %w", err)
+		reportFatalErr(logger, reporter, wrappedErr)
+		return wrappedErr
+	}
 
 	// Create Executor Factory
 	resMapping, err := execute.LoadResourceMappingIfShould(
 		logger.WithField(componentLogFieldKey, "Resource Mapping Loader"),
 		conf,
 		discoveryCli,
 	)
-	exitOnError(err, "while loading resource mapping")
+	if err != nil {
+		wrappedErr := fmt.Errorf("while loading resource mapping: %w", err)
+		reportFatalErr(logger, reporter, wrappedErr)
```

Comment:
I thought about doing `reportFatalErr` as deferred function (catch the returned error, and if it is not nil, report it). But that would require named output arguments, and also, that would be another deferred call (which should be executed before closing the reader). 🤔 On the other hand now we need to cover all exit paths with `reportFatalErr`. Let me know what do you think 🙂 I can change that to deferred calls if you'd like to.